### ISSUE TITLE
PhET-iO design review comments - Phase Changes Screen

### ISSUE THREAD

**Issue by arouinfar** [phetsims/states-of-matter#265]
For #245, and a continuation of #257.

- [x] `phaseChangesScreen.model.interactionStrengthProperty` - Add documentation to make it clear that this only applies to the adjustable attraction option. Does this value have units?

- [x] `phaseChangesScreen.model.numberOfMoleculesProperty` does seem to be hooked up properly. It is responsive to the BicyclePumpNode, but doesn't subtract molecules when the container explodes. (Particles lost when the container explodes should also refill the status bar on the BicyclePumpNode.)

- [x] `phaseChangesScreen.model.pressureProperty` needs units.

- [x] `view.compositeThermometerNode` - Add linked property to `model.temperatureInKelvinProperty`. (States & Phase Changes screens)

`phaseChangesScreen.view.pressureMeter`:
- [x] Rename pressureGaugeNode. 
- [x] Value readout needs to be instrumented. (Some clients may want to hide this.)
- [x] Add linked property to `model.pressureProperty` in value readout (or at the top-level of the pressureGaugeNode if this is too complicated).
- [x] Tree structure looks odd, gaugeNode doesn't need to be separately instrumented. Follow the pattern used in Gas Props, see `gasProperties.idealScreen.view.pressureGaugeNode`.  (Understood that there will be some differences since that sim has a ComboBox readout.)

The lid needs some restructuring to make related things easier to find:
- [x] Instrument the HandleNode. (If lid is not pickable, we want to be able to hide the handle.)
- [x] Create a container (maybe called LidNode, if appropriate) that contains `view.lidDragListener`, `view.pointingHandNode` and the HandleNode (currently not instrumented).

- [x] `phaseChangesScreen.view.returnLidButton` - set the enabledProperty, opacityProperty, pickableProperty, and visibleProperty to phetioReadOnly: true. (Similar to https://github.com/phetsims/gravity-and-orbits/issues/313).

- [x] Group the epsilon-related dragHandlers (epsilonResizeDragHandler and epsilonLineDragHandler) under a common parent called epsilonControls (or epsilonControlNode if more appropriate) that has top-level opacityProperty, pickableProperty, and visibleProperty.

AccordionBoxes (epsilonControlInteractionPotentialDiagram and phaseDiagram) need some restructuring. Currently, they look more like panels that contain an accordionBox.
- [x] Follow the pattern used in Gas Properties, e.g. `gasProperties.idealScreen.view.particlesAccordionBox`. There, the entire thing is an AccordionBox, rather than a container which has an AccordionBox inside.
- [x] Titles should be {{title}}AccordionBox: interactionPotentialAccordionBox & phaseDiagramAccordionBox
- [x] The corresponding ExpandedProperties in the model should be named {{title}}ExpandedProperty. Remove "Diagram" from `model.interactionPotentialDiagramExpandedProperty`.

**Comment by jbphet** [comment]
@arouinfar asked above:

- `phaseChangesScreen.model.interactionStrengthProperty`... Does this value have units?

Yes, at least kind of.  This value is epsilon in the Lennard-Jones potential equation, and epsilon/k-Boltzmann is in Kelvin.  So, I suppose that means the units for epsilon are k-Boltzmann-Kelvins, but that doesn't seem like it would be very meaningful to our users.  @arouinfar - shall I add these units or just leave them out?

**Comment by jbphet** [comment]
> `phaseChangesScreen.model.numberOfMoleculesProperty` does seem to be hooked up properly. It is responsive to the BicyclePumpNode, but doesn't subtract molecules when the container explodes. (Particles lost when the container explodes should also refill the status bar on the BicyclePumpNode.)

In the model, this value represents the number of particles being simulated, *not* the number in the container.  That means the number can't be updated until the lid is returned.  However, that wasn't happening either.  I've added code to update this value when the lid is returned, on resets, and on substance changes.  @arouinfar - please check the behavior on master and let me know if this is sufficient.

**Comment by arouinfar** [comment]
>> @arouinfar asked above:
>>
>>phaseChangesScreen.model.interactionStrengthProperty... Does this value have units?
>
>Yes, at least kind of. This value is epsilon in the Lennard-Jones potential equation, and epsilon/k-Boltzmann is in Kelvin. So, I suppose that means the units for epsilon are k-Boltzmann-Kelvins, but that doesn't seem like it would be very meaningful to our users. @arouinfar - shall I add these units or just leave them out?

Thanks for the clarification. Let's leave out the units. Can you adjust the documentation to say that it is a parameter in the Lennard-Jones potential?

>>phaseChangesScreen.model.numberOfMoleculesProperty does seem to be hooked up properly. It is responsive to the BicyclePumpNode, but doesn't subtract molecules when the container explodes. (Particles lost when the container explodes should also refill the status bar on the BicyclePumpNode.)
>
>In the model, this value represents the number of particles being simulated, not the number in the container. That means the number can't be updated until the lid is returned. However, that wasn't happening either. I've added code to update this value when the lid is returned, on resets, and on substance changes. @arouinfar - please check the behavior on master and let me know if this is sufficient.

Understood. The behavior looks good in master. Can you update the documentation its documentation to: This value represents the number of particles being simulated, not the number or particles in the container.

**Comment by jbphet** [comment]
From https://github.com/phetsims/states-of-matter/issues/265#issue-590371300 above:

> Tree structure looks odd, gaugeNode doesn't need to be separately instrumented. Follow the pattern used in Gas Props...

@arouinfar - This would actually be fairly hard to restructure, since the node consists of the gauge, the readout, the collar, and the connector pipe.  It seems odd to me that the readout would be separated but the gauge wouldn't.  Are you sure on this one?  Would it be more clear if I were to instrument the collar and the pipe?

**Comment by jbphet** [comment]
I think I've addressed everything in the list above, with some caveats noted in subsequent comments.  @arouinfar - can you please review and make sure that I've truly caught everything?  If not, please note any discrepancies and assign back to me for another round.

**Comment by arouinfar** [comment]
@jbphet a few more tweaks, but some of them should be fairly easy.

- [ ] `phaseChangesScreen.model.numberOfMoleculesProperty`: There was a typo in my recommended documentation that got pasted into the sim. It currently reads "number or particles" but should be "number of particles".

- [ ] `phaseChangesScreen.model.interactionPotentialAccordionBoxExpandedProperty`: The "AccordionBox" doesn't need to be part of tandem name. (Sorry, my original request could have been worded better.) Please change it to `interactionPotentialExpandedProperty`.

- [ ] Likewise, remove the "AccordionBox" from `phaseChangesScreen.model.phaseDiagramAccordionBoxExpandedProperty` and change it to `phaseDiagramExpandedProperty`.

- [ ] `phaseChangesScreen.view.interactionPotentialAccordionBox.diagram.positionMarker`: This is irrelevant to this screen. Does it need to be here? Can it be removed?

- [ ] The title of the Interaction Strength slider in the moleculesControlPanel need to be instrumented. It's possible to hide the slider, but not the title. Please follow the pattern used in `interactionsScreen.view.atomsControlPanel.interactionStrengthSlider`. (I forgot to request the change on this screen the first time around, sorry!)

We like the direction of the ParticleContainerNode, but think it could be further improved. It seems odd that the pressure gauge is part of the container, but the thermometer is not. In the current implementation, clients must make the entire container unpickable to prevent users from moving the lid (no other way to disable the lidDragListener). This isn't particularly intuitive, and would also mean making the thermometer's ComboBox's unpickable. This was the tree structure we discussed in yesterday's zoom call. 

```
-particleContainerNode
     opacityProperty
     pickableProperty
     visibleProperty
     +compositeThermometerNode
     -lidNode
           opacityProperty
           pickableProperty
           visibleProperty
           +handleNode
           +lidDragListener
           +pointingHandNode
     +pressureGaugeNode
```

- [ ] Move `compositeThermometerNode` into the `particleContainerNode`.
- [ ] Create a `lidNode` to contain the handleNode, lidDragListener, and pointingHandNode. The goal is for clients to be able to disable the lid without needing to talk to the individual parts.
- [ ] Make `particleContainerNode.lidNode.visibleProperty` phetioReadOnly: true.

**Comment by arouinfar** [comment]
We also discussed how to handle the epsilonControls in the Interaction screen, and the decisions we make in #266 will likely affect the tree structure of the interaction potential diagram, which currently looks like this.

![image](https://user-images.githubusercontent.com/8419308/79493303-88094e80-7fde-11ea-9289-8e03da26b557.png)

Re-assigning myself to make sure update this issue once things have been decided in #266.

**Comment by arouinfar** [comment]
Based on the tree structure proposed in https://github.com/phetsims/states-of-matter/issues/266#issuecomment-614936661, we'd like to update the interactionPotentialAccordionBox:

```
-interactionPotentialAccordionBox
     opacityProperty
     pickableProperty
     visibleProperty
     -epsilonParameter
         -controlArrow
               opacityProperty
               pickableProperty
               visibleProperty
              +dragListener 
         -controlLine
               opacityProperty
               pickableProperty
               visibleProperty
               +dragListener
          -graphLabel (includes double headed white arrow & epsilon label)
               opacityProperty
               pickableProperty
               visibleProperty
     +expandCollapseButton
     -sigmaParameter
         -graphLabel (includes double headed white arrow & sigma label)
               opacityProperty
               pickableProperty
               visibleProperty
```

@jbphet we removed the "diagram" level from the AccordionBox. If we need to keep the container, the we recommend calling it "graph" to be consistent with how we'd like to name things on the Interaction screen.

**Comment by jbphet** [comment]
@arouinfar - in response to [this comment](https://github.com/phetsims/states-of-matter/issues/265#issuecomment-614942981) above, I'm not able to pull out the `diagram` level in a reasonable way because it really is a separate part of the accordion box, and it would take some odd tandem manipulations that seem inappropriate to me.  I have changed the name of that level to `graph` as suggested.

Also, I can't put the epsilon and sigma labels together with the controls because they are created in completely different parts of the code, so I composited the labels (one each for the epsilon and sigma parameters) and added them at the `graph` level.  This makes the names seem a little odd, and we may want to change `epsilonParameter` to `epsilonControl` and just have `arrow` and `line` underneath, but I left it in a state where it matches your originally suggested names pretty closely.  I'll leave it up to you as to whether you'd like to make some updates to the names.

Other than that I was able to get reasonably close to what was requested.  Please review and let me know what else should be changed.

**Comment by jbphet** [comment]
I think I've now addressed everything from the most recent review notes.  @arouinfar - back to you for review.

**Comment by arouinfar** [comment]
> The title of the Interaction Strength slider in the moleculesControlPanel need to be instrumented. It's possible to hide the slider, but not the title. Please follow the pattern used in interactionsScreen.view.atomsControlPanel.interactionStrengthSlider. (I forgot to request the change on this screen the first time around, sorry!)

Something's not quite right here. If I set
`phaseChangesScreen.view.moleculesControlPanel.interactionStrengthSlider.visibleProperty` to false, the "Interaction Strength" title is still visible, even though it's nested under the slider:
`phaseChangesScreen.view.moleculesControlPanel.interactionStrengthSlider.title.visibleProperty`

The updated lidNode is looking really nice! I realize that `lidEllipseNode` probably needed to be instrumented for everything to work, but I think it looks really strange if it's made invisible. The lid appears to be gone, but it's still in the model and particles can collide with it. Can you make its visibleProperty phetioReadOnly: true. (We did the same for lidNode.visibleProperty.)
`phaseChangesScreen.view.particleContainerNode.lidNode.lidEllipseNode.visibleProperty`

The contents of the interactionPotentialAccordionBox are looking really nice with all of the updates. We'd really like it to be possible for clients to permanently hide the epsilonControls. Toggling the visibleProperties of the arrow and line to false do not permanently remove these controls. Each time the user selects "Adjustable Attraction" these will be made visible again. 
`phaseChangesScreen.view.interactionPotentialAccordionBox.graph.epsilonControl.arrow.visibleProperty`
`phaseChangesScreen.view.interactionPotentialAccordionBox.graph.epsilonControl.line.visibleProperty`

We faced a similar challenge with the NeutralIndicatorNode in ph-scale, see https://github.com/phetsims/ph-scale/issues/102.

@jbphet can you investigate and see if this approach would be applicable to SOM?

**Comment by arouinfar** [comment]
@jbphet I discovered that I can make the epsilonControl.arrow and epsilonControl.line visible for a non-Adjustable Attraction atom. This looks super buggy. For example:

![image](https://user-images.githubusercontent.com/8419308/80833876-c4c37100-8bac-11ea-8289-c720cd71ef0c.png)

**Comment by jbphet** [comment]
Here's a screenshot of the structure that @arouinfar and I came up with today that we'd like to see for supporting hiding of the interactive controls in the interaction potential graphs.  Note the capitalization is off - it's due to the word processor trying to be helpful.

![image](https://user-images.githubusercontent.com/3082831/81452191-73cef200-9143-11ea-9305-464b796bc4d8.png)

**Comment by jbphet** [comment]
@arouinfar - I believe I've made the updates requested in [this comment above](https://github.com/phetsims/states-of-matter/issues/265#issuecomment-622519379).  Please check it out on master and let me know what you think.

I also added some documentation to the interaction controls as we discussed, and I used the phrase "Used for Adjustable Attraction only".  I'm happy to change it, I just wanted to get something in there for now.

**Comment by arouinfar** [comment]
@jbphet are the opacityProperty, pickableProperty, and visibleProperty of epsilonArrow, epsilonLine, and sigmaArrow intentionally phetioReadOnly: true? We had talked about clients potentially wanting to hide just the epsilon-related or sigma-related graph controls.

Everything else from https://github.com/phetsims/states-of-matter/issues/265#issuecomment-622519379 looks great.

**Comment by jbphet** [comment]
> @jbphet are the opacityProperty, pickableProperty, and visibleProperty of epsilonArrow, epsilonLine, and sigmaArrow intentionally phetioReadOnly: true? We had talked about clients potentially wanting to hide just the epsilon-related or sigma-related graph controls.

@arouinfar - I guess I came away with a different understanding of what we had talked about.  Because the visibility attributes of these nodes are what is used to hide and show them when the user moves back and forth between adjustable and non-adjustable atoms, we wanted the attributes to be read-only so that the user couldn't create problematic states, such as the one you pointed out in https://github.com/phetsims/states-of-matter/issues/265#issuecomment-622521857, and also so that they don't try to hide the controls only to be thwarted by switching atoms and having them come back.

If you believe that it's important the the user be able to separately hide the epsilon and sigma controls, I would recommend inserting another layer in the hierarchy shown in https://github.com/phetsims/states-of-matter/issues/265#issuecomment-626036101 (and currently in master).

**Comment by arouinfar** [comment]
@jbphet sorry for the miscommunication. I didn't realize the epsilonArrow, epsilonLine, and sigmaArrow in the tree hierarchy in https://github.com/phetsims/states-of-matter/issues/265#issuecomment-626036101 were still going to be the actual model parameters that would be overwritten when switching atoms.

For the Phase Changes screen (which only has an epsilon control) it is sufficient to just toggle `view.interactionPotentialAccordionBox.graph.interactiveControls.visibleProperty`to get rid of the controls in the graph. I don't think we need to make any changes.

For the Interaction screen, we'll need the extra layer in the tree to hide the sigma and epsilon controls separately. (The top-level `interactiveControls.visibleProperty` is still useful for clients who may want to hide both controls, so leave that as-is.)

**Comment by arouinfar** [comment]
Discovered in #272 

- [ ] Toggling `phaseChangesScreen.view.moleculesControlPanel.radioButtonGroup.visibleProperty` should only affect the visibility of the radioButtonGroup. Currently, the panel stroke disappears which is undesirable (and inconsistent with the equivalent behavior on the statesScreen).

**Comment by arouinfar** [comment]
@jbphet here's an outline of the tree structure we just discussed

```
interactivePotentialGraph
    -interactiveControls
        opacityProperty
        pickableProperty
        visibleProperty
        -epsilon		
             opacityProperty
             pickableProperty
             visibleProperty
             -epsilonArrow
                   opacityProperty
                   pickableProperty
                   visibleProperty
             -epsilonLine
                   opacityProperty
                   pickableProperty
                   visibleProperty
        -sigma		
             opacityProperty
             pickableProperty
             visibleProperty
             -sigmaArrow
                   opacityProperty
                   pickableProperty
                   visibleProperty
```

**Comment by jbphet** [comment]
> Toggling phaseChangesScreen.view.moleculesControlPanel.radioButtonGroup.visibleProperty should only affect the visibility of the radioButtonGroup. Currently, the panel stroke disappears which is undesirable (and inconsistent with the equivalent behavior on the statesScreen).

Some changes have been make recently to `Panel.js` such that it automatically does a re-layout when elements are added or removed from the panel.  This is done primarily in support of PhET-iO.  A decision was made that, if everything is removed from the panel, the panel itself should disappear.  I conveyed this to @arouinfar during a Zoom call, and (if I recall correctly) it was mostly the inconsistency between screens that she was concerned about.

I showed the behavior to @jonathanolson and he made changes that makes the panel's behavior more consistent, so now both panels disappear if the contents (i.e. the radio button groups) are made invisible.  I've also make changes that cause the title to disappear in this case.  Assigning back to @arouinfar to see if this behavior, as well as the restructured tree for the interactive sigma & epsilon controls, look like what we're after.
